pubs/bibstruct.py

```diff
@@ -56,26 +65,51 @@ def valid_citekey(citekey):
     # FIXME: a bit crude, but efficient for now (and allows unicode citekeys)
     return not '/' in citekey
 
-
-def generate_citekey(bibdata):
+class CitekeyFormatter(Formatter):
+    def __init__(self):
+        super(CitekeyFormatter, self).__init__()
+
+    def format_field(self, val, fmt):
+        if len(fmt) > 0 and fmt[0] == 'u':
+            s = str(val).upper()
+            fmt = fmt[1:]
+        elif len(fmt) > 0 and fmt[0] == 'l':
+            s = str(val).lower()
+            fmt = fmt[1:]
+        else:
+            s = val
+        return str2citekey(s.__format__(fmt))
+
+
+    def get_value(self, key, args, entry):
+        if isinstance(key, (str, unicode)):
+            okey = key
```

Comment:
Being less terse is useful here: `original_key` instead of `okey` would help the next programmer understand faster.